Cleanup dataframes

nkanazawa1989 opened the pull request:
### Summary

This PR updates the implementation of `ScatterTable` and `AnalysisResultTable` based on the [comment](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1319#issuecomment-1827343405) from @itoko . 

### Details and comments

Current pattern heavily uses inheritance; `Table(DataFrame, MixIn)`, but this causes several problems. Qiskit Experiments class directly depends on the third party library, resulting in Sphinx directive mismatch and poor robustness of the API. Instead of using inheritance, these classes are refactored with composition and delegation, namely
```python
class Table:
    def __init__(self):
        self._data = DataFrame(...)
```
this pattern is also common in other software libraries using dataframe. Since this PR removes unreleased public classes, this should be merged before the release.

Although this updates many files, these are just delegation of data handling logic to the class itself, which simplifies the implantation of classes that operate the container objects. Also new pattern allows more strict dtype management with dataframe.

wshanks commented:
This is good timing because the `ScatterTable` inheritance model just broke with the release of pandas 2.2. See for example the test failures on #1363 [here](https://github.com/Qiskit-Extensions/qiskit-experiments/actions/runs/7605764016/job/20710510778?pr=1363).

nkanazawa1989 commented on qiskit_experiments/library/driven_freq_tuning/coefficients.py:
```diff
@@ -181,6 +181,7 @@ def __eq__(self, other):
                 self.neg_coef_o1 == other.neg_coef_o1,
                 self.neg_coef_o2 == other.neg_coef_o2,
                 self.neg_coef_o3 == other.neg_coef_o3,
+                self.offset == other.offset,
```
This is a followup of #1236 . PR can be separated.

wshanks replied:
One of the nice things about dataclasses :slightly_smiling_face:

coruscating reviewed (commented):
`auto_save` is not working right now. I get errors like `{"errors":["experiment_uuid is required","type must be a string","chisq is not a decimal"]}` after this:

> expdata=exp.run(backend)
> expdata.auto_save=True

Then if I run `save()` again on its own, everything seems to save correctly.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,334 @@
+        return instance
+
+    @property
+    def dataframe(self):
```
One concern for returning the html representation is that a user might believe the object is a bare pandas dataframe and blame missing methods that pandas provides. Of course html representation is very helpful, and I'm not against adding it.

coruscating replied:
That's true, maybe there can be a header before the dataframe content to distinguish the table from a dataframe.

nkanazawa1989 commented:
Thanks @coruscating for catching the bug. The experiment ID is not recorded in the analysis result table (because the experiment name and ID are sort of duplication and only human-friendly name is stored) and the service AnalysisResult object created from the table requires the ID to look for where to save the data. Likely this was introduced in the previous PR. Fixed in [901bb28](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/901bb281b00bba3060e7c8746ce467d1d75981a9)

itoko commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,434 @@
+    def x(self, new_values):
+        self._data.loc[:, "xval"] = new_values
+
+    def get_x(
```
How about `xvals`? same characters but suggesting an array will be returned.

nkanazawa1989 replied:
Done in [92cfc92](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/92cfc92f5e0be9556155e25a1152a1c9abfdecf5)

itoko commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,434 @@
+        self._data.loc[:, "name"] = new_values
+
+    @property
+    def class_id(self) -> np.ndarray:
```
Any better name? `data_uid`?

nkanazawa1989 replied:
Done in [92cfc92](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/92cfc92f5e0be9556155e25a1152a1c9abfdecf5)

itoko reviewed (commented):
@nkanazawa1989 , thank you for your hard work on this cleanup. I've mainly reviewed public interface of `ScatterTable` class newly introduced during this series of refactoring. Overall, looks good to me in that point. 

Just one major comment from me. To tell the following points clearly to users, I think we need more introductory explanation is necessary at the beginning of `ScatterTable` class (and/or in `docs/tutorials/curve_analysis.rst`). For example, we may need to explain the important tuple (`kind`, `category`, `analysis`) as a unique key; We may need to mention to what values are allowed for `category` and what each of them represents; so on.

I understand `ScatterTable` object now to be a single source of the truth as the data used for plotting. I think that's good for users who rework with some complex Analysis classes. Instead, it may contain multiple data sets. That requires users to fully understand the three concepts, `kind`, `category` and `analysis`, because this triplet must be a unique key to retrieve one sequence of data points (x, y, y_err) from multiple data sets. I think that's why all of the versatile `filter` method and `get_?` methods have those in their options. I personally like `get_?` methods since most of experiments would be simple and `ScatterTable` object is likely to have a single key (`kind`, `category` and `analysis`) for all rows. They would prevent from calling `filter` functions everywhere.

I'll left minor comments inline, each of which should not prevent merging this PR.

nkanazawa1989 commented:
Thanks @itoko , I added more documentation in [346d23a](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/346d23af58bee40fa005e030dadb7ca9f08a7470). I added the documentation directly to the class doc of the ScatterTable for visibility and added the cross-ref link to the tutorial.  

> most of experiments would be simple and ScatterTable object is likely to have a single key (kind, category and analysis) for all rows

This is not true because even the model and analysis are trivial most of analysis subclasses produce two categories of raw and formatted, e.g. in RB, raw corresponds to P1 of each randomization seed and formatted is P1 of the sample average of them. I hesitate to add dedicated methods like `raw_xvals` and `formatted_xvals` for shortcut because naming rule depends on analysis; e.g. [StarkRamseyXYAmpScanAnalysis](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/b4fc262eb828f838e28d7a354428d0213a90465d/qiskit_experiments/library/driven_freq_tuning/ramsey_amp_scan_analysis.py#L253) has four columns "raw", "ramsey_xy", "freq", "fitted".

coruscating commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,497 @@
-            (table.name == "sub_exp_1") & (table.category == "formatted")
-        ]
-        x, y, e = abc_data.xval.to_numpy(), abc_data.yval.to_numpy(), abc_data.yerr.to_numpy()
+        mini_table = table.filter(kind="model1", category="raw", analysis="Analysis_A")
```
I find using `kind` as a filter name a bit confusing since it means the same as category. How about `model` or `class` instead?

nkanazawa1989 replied:
Replaced with `data_uid`. Usually it matches with the model but some case not. For example, in the StarkRamseyXYAmpScanAnalysis in the raw category this field represents either X or Y measurement data but after formatted the field represents frequency data in the fit model. class is reserved word.

coruscating reviewed (commented):
Thanks @nkanazawa1989, I like the lazy add rows approach and the overall the code looks fine to me. Since it's difficult to test the ScatterTable output from experiments in this PR, perhaps we can merge this first and handle any potential followups from adding unit tests in #1342 in another PR, if needed.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,394 @@
-    """
+    def __init__(self):
+        """Create new dataset."""
+        super().__init__()
```
good catch! thanks

nkanazawa1989 commented on docs/tutorials/curve_analysis.rst:
```diff
@@ -319,12 +320,12 @@ This may return new scatter table object with the addition of rows like the foll
-    15   0.1  0.483415  0.010774    B         1  formatted   2048
-    16   0.2  0.426829  0.010678    B         1  formatted   2048
-    17   0.3  0.568293  0.008592    B         1  formatted   2048
+    12   0.1  0.234634  0.009183    A         0  formatted   2048   MyAnalysis
```
Yes, done in [ee03161](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/ee0316148645f25acf57e7ba164da9dfd7ef8012)

wshanks commented on qiskit_experiments/framework/analysis_result_table.py:
```diff
@@ -10,256 +10,317 @@
+                    "quality",
+                ] + extra_columns
+            if not isinstance(columns, str):
+                out = []
```
I ran the tests with coverage to see how well tested ScatterTable and AnalysisResultTable were. Besides handling errors, this block for `get_data` with a list of column names was the only significant case not covered.

nkanazawa1989 replied:
Thanks for the test coverage check. Added more tests in [ee92f1d](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/ee92f1df501db02c34505b52581d558cee964bae)

wshanks reviewed (commented):
I read through all the changes and left comments. There are many but I think they are easy to address and most are just related to documentation.

The thing I am most uncertain about with ScatterTable is how much it is a user facing class versus an internal class. For a user just running experiments, I think it is not exposed, but for a user writing CurveAnalysis subclasses it is? So it only makes sense to document it in the context of curve analysis? There is a little documentation now but not too much (the table is shown in the the curve analysis workflow section and the columns are described in the ScatterTable doc string). One concern I have to make the roles of `name`, `data_uid`, `category`, and `model` (not a column but a term used in the same context) clear.

wshanks commented:
Also, I read all the code changes but one thing that is hard with a PR like this is to review all the code that didn't change. I tried checking the documentation and release notes and I think they still make sense with the changes that were made.

itoko commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
+    and data points can also take metadata triplet (`data_uid`, `category`, `analysis`)
+    to distinguish the subset.
+
+    * The `data_uid` is an integer key representing the class of the data.
```
Thank you for adding this info in the doc, @nkanazawa1989 . It helps me a lot to understand the concepts to know for users of ScatterTable class. I've now understood `data_uid` identifies a "series" of data to fit a model. And `name` is another human-friendly identifier that indicates a "series" of data to fit a model. Given that, `series_id` and `series_name` sound better to me. What do you think? @nkanazawa1989 @wshanks

itoko replied:
`series_index` instead of `series_id`? Looking at the example in the tutorial, `data_uid` can take a common value for different category. In that sense, it's not "ID" (to me, ID sounds like a solo unique identifier), it's more like "index" defined for each (`category`, `analysis`) data.

coruscating replied:
I like this suggestion. Perhaps `series_idx` to be shorter?

wshanks replied:
One case to consider regarding the naming is `StarkRamseyXYAmpScanAnalysis`. For that experiment, X and Y vs amp data are recorded. Then the X and Y data are converted into phase vs amp data which is fit. So there are two raw series that become the formatted series that is fit. Maybe `series_` still makes sense since one series is passed to the model. It is the fit models though that have a one to one mapping the `data_uid` values.

itoko replied:
Umm, I'm converging to Will's comment:
> One concern I have to make the roles of name, data_uid, category, and model (not a column but a term used in the same context) clear.

Maybe, all we need is `model_` (we can distinguish "ramsey_xy" and "freq" using `model_name` instead of `category`).

nkanazawa1989 replied:
We cannot define fit model for ramsey_xy because its P1 envelope is unpredictable. In this sense, "ramsey_xy" is another the intermediate data to produce the frequency curves which have fit models (Ramsey data is used for visualization and we cannot discard -- when fit doesn't go well experimentalists can get some insights from the raw Ramsey signal). This is why I cannot introduce fixed categories -- this requires the Ramsey analysis to define ScatterTable subclass with its own categories.

nkanazawa1989 replied:
Anyways I also like series. This was originally used in the curve analysis to indicate the data belongs to a particular fit model.

nkanazawa1989 replied:
Done in [f9d13bd](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/f9d13bd59b27deab0e49a0f535dde3375bb051b0). Thanks for the suggestion.

wshanks replied:
So I think the broadest way to think about the columns are that `series` labels a single curve of some kind (`ScatterTable` is under `qiskit_experiments.curve_analysis` so it is safe to assume all the data belong to curves). `category` is a label for a group of curves/series that all belong to the same stage of processing (raw/formatted/fitted). `analysis` is a namespace for separating these groups of curves from different subexperiments.

@coruscating Is it important to get the release out today? I think I would like to clarify these terms a bit in the docs, but that could be done in a follow up if it is important to release today. I don't have any objections to the code (as long as further discussion doesn't make us want to change any names again).

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
+    Each data point in a 1-D curve fit may consist of the x value, y value, and
+    standard error of the y value.
+    In addition, such analysis may internally create several data subsets,
+    and data points can also take metadata triplet (`data_uid`, `category`, `analysis`)
```
Yes, these don't indicate code and I think italic is a reasonable decoration here.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
+    and data points can also take metadata triplet (`data_uid`, `category`, `analysis`)
+    to distinguish the subset.
+
+    * The `data_uid` is an integer key representing the class of the data.
```
Added more description in [9e27f16](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/9e27f16eb3d0f376cc0bf81494319199d67f4cb1)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
+
+    .. _filter_scatter_table:
+
+    Filtering ScatterTable
```
Fair enough. I wrote the docs here just for visibility (since developer may not check website tutorial) but this makes rendered doc bit ugly. I moved this section to the tutorial and added the cross ref link here. Done in [9e27f16](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/9e27f16eb3d0f376cc0bf81494319199d67f4cb1)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
+    @property
+    def dataframe(self):
+        """Dataframe object of data points."""
+        return self._data
```
You're right. The protected property method is not necessary. Removed in [b870be3](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/b870be33607b502dc5e961a40b2e8c1785fec05c)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
-        return self[self.name == index]
+        if list(data.columns) != cls.DEFAULT_COLUMNS:
+            raise ValueError("Input dataframe columns don't match with the ScatterTable spec.")
+        instance = object.__new__(ScatterTable)
```
This is to avoid creation of unnecessary dataframe in the constructor. Alternatively we can do `def __init__(self, data: pd.DataFrame):` instead of the factory method.

wshanks replied:
Okay, hmm, maybe an `__init__` argument would be cleaner though I don't have a strong preference. With the current way `from_dataframe` needs to be kept in sync with the behavior of `__init__`. `from_dataframe` is used in several tests, so we should catch if something is added to `__init__` but not here.

nkanazawa1989 commented on qiskit_experiments/framework/analysis_result_table.py:
```diff
@@ -10,256 +10,317 @@
+    def dataframe(self) -> pd.DataFrame:
+        """Dataframe object of analysis results."""
+        with self._lock:
+            return self._data.copy(deep=False)
```
Yes, that's the intention. I think shallow copy is the tradeoff between thread-safety and performance (but this is mainly end-user facing API and probably deep-copy is okey).

nkanazawa1989 commented on qiskit_experiments/framework/analysis_result_table.py:
```diff
@@ -10,256 +10,317 @@
+            else:
+                key = self._create_unique_hash()
+        if data.get("result_id", None) is None:
+            data["result_id"] = key
```
I realized that the key argument is not necessary. This code comes from my experimental branch that tries to introduce a protocol for the data containers, e.g. figures, circuit results, analysis results, artifacts, etc.... Removed and code is refactored in [cc905c6](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/cc905c689ac7f51ff0eae4c19afe81935956114f)

nkanazawa1989 commented on qiskit_experiments/framework/analysis_result_table.py:
```diff
@@ -10,256 +10,317 @@
+                if key in self._data.index:
+                    return [key]
+                # This key is name of entry
+                loc = self._data["name"] == key
```
Yeah unfortunately this is existing behavior :(

nkanazawa1989 commented on qiskit_experiments/framework/experiment_data.py:
```diff
@@ -1569,30 +1557,19 @@ def analysis_results(
-            out = out[valid_columns]
-            if len(out) == 1 and index is not None:
+            df = self._analysis_results.get_data(index, columns=columns)
+            if len(df) == 1 and index is not None:
```
Yes this makes sense. My original concern was the our user's degree of familiarity with the dataframe API (especially `.iloc[0]` is bit tricky). But I agree this typecase introduces extra code overhead everywhere, and decided to remove with sufficient documentation. See [0dc4eb2](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/0dc4eb299a6691ca0849f087893eb430d4a816f6)

nkanazawa1989 commented on qiskit_experiments/curve_analysis/scatter_table.py:
```diff
@@ -9,185 +9,498 @@
         """X values."""
-        return self.xval.to_numpy()
+        # For backward compatibility with CurveData.x
+        return self._data.xval.to_numpy(dtype=float, na_value=np.nan)
```
IIRC using the lru_cache for the class method may cause memory leak and I guess I need to implement custom logic (i.e. cache that clears the data when `_lazy_add_rows` is changed). 

According to the performance analysis of T1Analysis, the most time-consuming logic after [03aac67](https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1360/commits/03aac67c30c5e74fa2b07047d888a6937292c65f) is still `iter_groups`, and the most expensive part is the data extraction from the dataframe. This means implementation of the cache doesn't drastically improve the performance.
![image](https://github.com/Qiskit-Extensions/qiskit-experiments/assets/39517270/39e17a4a-d108-44f9-a2ed-0cb1011ac11c)

Probably we want to replace pandas with polars in the next release.

wshanks replied:
Yes, I was going to comment on `iter_groups` that we should look at using polars there. Even just casting to polars for that one method and back to pandas might help (I mean, not refactoring the rest of the code to use polars).

nkanazawa1989 replied:
Another option would be to stop over-engineering. We currently allow users to run experiment on existing experiment data to get more statistically confident result. But I don't think this is heavy used feature and we can remove. In principle we run very expensive operation for the sake of something useless; if we have 100 xvals, we need to iterate over them and create new table with size=1 to average... Even if we switch to polars there still be overhead of object creation.

wshanks replied:
Yes, I agree that that feature has never seemed that useful to me. I think we could implement some support for post-processing instead, like a `combine_results` function or something that could take multiple experiment results, warn if the options are not all the same, and then run the analysis with the combined job data.

We didn't have this ability to extend the data of a completed experient in the code we used before qiskit-experiments and I think it was a feature request of something to be included when starting over from scratch. I like the experiment being less mutable and extending the experiment being like "copy-on-write" where a new experiment ID is generated for the combination of multiple experiments if you need that.

nkanazawa1989 replied:
I like your suggestion. IIRC when we started this project we had QV experiment in our mind. Namely we need to repeat until standard error gets converged within 1 sigma. The rest of experiments don't require such mechanism. Combine results is something useful :)

nkanazawa1989 replied:
Sorry this feature was already removed by https://github.com/Qiskit-Extensions/qiskit-experiments/pull/463. This means we can now just drop this formatter and raise a warning if we encounter duplicated xvals -- bur likely for next release.

wshanks replied:
Okay, I opened #1394 to try to capture what we said here but maybe its description is out of date.

nkanazawa1989 commented on qiskit_experiments/curve_analysis/composite_curve_analysis.py:
```diff
@@ -376,32 +377,35 @@ def _run_analysis(
+                        yerr_arr_fit = unp.std_devs(uval_arr_fit)
+                    else:
+                        yerr_arr_fit = np.zeros_like(xval_arr_fit)
+                    for xval, yval, yerr in zip(xval_arr_fit, yval_arr_fit, yerr_arr_fit):
```
Handling of the empty column is expensive because it requires careful handling of missing values. Without doing this shots column may be accidentally typecasted to float because numpy doesn't support nullable integer. This means we first need to create a 2D object-dtype ndarray and populate values, then convert it into dataframe. Since current `_lazy_add_rows` buffer assumes row-wise data list, arrays needs to be converted into this form internally.

nkanazawa1989 commented on docs/tutorials/curve_analysis.rst:
```diff
@@ -240,6 +240,85 @@ generate initial guesses for parameters, from the ``AnalysisA`` class in the fir
+- `shots`: Number of measurement shots used to acquire a data point. This value can be defined in the circuit metadata.
+- `analysis`: The name of the curve analysis instance that generated a data point.
+
+This object helps an analysis developer with writing a custom analysis class
+without an overhead of complex data management, as well as end-users with
+retrieving and reusing the intermediate data for their custom fitting workflow
+outside our curve fitting framework.
+Note that a :class:`ScatterTable` instance may be saved in the :class:`.ExperimentData` as an artifact.
```
I'll ask @coruscating to add cross-ref to new tutorial in #1342

coruscating replied:
Done in 79370a7.

wshanks commented on test/curve_analysis/test_scatter_table.py:
```diff
@@ -0,0 +1,270 @@
+        }
+        self.reference = pd.DataFrame.from_dict(source)
+
+    def test_create_table_from_dataframe(self):
```
Since from `from_dataframe` is an alternate constructor, could we do a couple other tests here? The ones I am thinking of are 1. use `self.reference.itertuples()` to iterate over the reference data and call `add_row` on an empty `ScatterTable` and then compare to `ScatterTable.from_dataframe(self.reference)` and 2. for those two scatter tables compare `.__dict__` or at least `.__dict__.keys()` to make sure the alternative constructor does not miss any private attributes from `__init__`.
